fix: NetworkAnimator not detecting and synchronizing cross fade initiated transitions

NoelStephensUnity opened the pull request:
This PR resolves the issue where `NetworkAnimator` was not properly detecting and synchronizing cross fade initiated transitions and not properly synchronizing changes to animation states.
It also reduces the cost of each animation state update by packing eligible properties.

## Changelog

- Fixed: Issue where `NetworkAnimator` was not properly detecting and synchronizing cross fade initiated transitions.
- Fixed: Issue where `NetworkAnimator` was not properly synchronizing animation state updates.

## Testing and Documentation

- Includes integration test.
- Includes updates to manual test.
- No documentation changes or additions were necessary.

NoelStephensUnity committed:
fix

For client authoritative, the server now actually sends any pending animation updates to any non-authoritative clients.

NoelStephensUnity committed:
fix

Trigger not updating properly on server when server authoritative but was the owner sending the trigger update.

NoelStephensUnity committed:
fix

Improve detection of the type of animation state change.

NoelStephensUnity committed:
fix

This fixes the issue where cross fades were not being handled by NetworkAnimator.
This is a first pass fix and the serialization could use an overhaul to reduce bandwidth costs.

NoelStephensUnity committed:
update

Making sure we don't toggle between transition and crossfade.
Packing values for AnimationState

NoelStephensUnity committed:
update

Updating the manual test used to get cross fading to AnimationStates without transitions working.

NoelStephensUnity committed:
update and style

Minor updates and cleanup
Added additional comments.

NoelStephensUnity committed:
test

Adding some cross fade tests to the NetworkAnimatorTests.
Also fixed an issue with the coroutineRunner being destroyed before the IntegrationTestSceneHandler was.

NoelStephensUnity committed:
test

This is the final test for cross fade initiated transitions and whether it is detected and synchronized or not.

NoelStephensUnity committed:
update

updating the changelog

ShadauxCat commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -317,77 +326,50 @@ internal struct AnimationState : INetworkSerializable
-                    writeSize += FastBufferWriter.GetWriteSize(Layer);
-                    writeSize += FastBufferWriter.GetWriteSize(Weight);
+                    BytePacker.WriteValuePacked(writer, Transition);
+                    BytePacker.WriteValuePacked(writer, CrossFade);
```
Just a note that packing on a bool or byte value is just a passthrough to `writer.WriteValueSafe(value)`. Not anything that needs changing, just wanted to note this in case you were expecting something different.

NoelStephensUnity replied:
Actually, I should make that a single byte with flags... thanks for the call out.  👍

NoelStephensUnity committed:
update

Making the serialization of the two bools one byte instead of two bytes.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-notsending-animationstates

NoelStephensUnity committed:
update

removing redundant if check for IsServerAuthoritative

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -317,77 +326,69 @@ internal struct AnimationState : INetworkSerializable
             {
-                if (serializer.IsWriter)
+                var isWriter = serializer.IsWriter;
+                if (isWriter)
```
Just curious but why make this change?

NoelStephensUnity replied:
Good catch!  I originally was going to be checking that at other locations but ended up serializing in a way that did not need that.  There is no reason for that adjustment and will be committing an update that removes it.  👍

samlucas-unity commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -317,77 +326,69 @@ internal struct AnimationState : INetworkSerializable
+                        m_StateFlags |= k_IsTransition;
                     }
-
-                    if (!writer.TryBeginWrite(writeSize))
```
This seemed to explicitly want to check it could write the data to error here. Where will it error now for the same condition?

NoelStephensUnity replied:
The primary reason for the switch was due to the added byte packing to reduce bandwidth costs per update (which the size depends upon how well the property is packed)... the remaining serialized properties didn't make sense, processing wise, to do that same pre-calculation so I just ran with safe serialization for everything so that issue is no longer of concern.

NoelStephensUnity committed:
update

removing extra local variable being used to determine if the serilization was in the write or read context.
